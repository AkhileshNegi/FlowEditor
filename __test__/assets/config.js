const config = {
    flow: 'boring'
};

module.exports =
    process.env.NODE_ENV === 'production'
        ? Object.assign({}, config, {
              localStorage: true,
              endpoints: {
                  flows: 'flows',
                  groups: 'groups',
                  contacts: 'contacts',
                  fields: 'fields',
                  labels: 'labels',
                  activity: '',
                  engine: ''
              }
          })
        : Object.assign({}, config, {
              localStorage: true,
              endpoints: {
                  flows: '/assets/flows',
                  groups: '/assets/groups',
                  contacts: '/assets/contacts',
                  fields: '/assets/fields',
                  labels: '/assets/labels',
                  activity: '',
                  engine: ''
              }
          });
