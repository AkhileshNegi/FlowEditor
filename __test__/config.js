const config = {
    flow: 'a4f64f1b-85bc-477e-b706-de313a022979'
};

module.exports =
    process.env.NODE_ENV === 'production'
        ? Object.assign({}, config, {
              localStorage: true,
              endpoints: {
                  flows: 'flows',
                  groups: 'groups',
                  recipients: 'recipients',
                  fields: 'fields',
                  labels: 'labels',
                  languages: 'languages',
                  channels: 'channels',
                  environment: 'environment',
                  activity: '',
                  simulateStart: '',
                  simulateResume: ''
              }
          })
        : Object.assign({}, config, {
              localStorage: true,
              endpoints: {
                  flows: '/assets/flows.json',
                  groups: '/assets/groups.json',
                  fields: '/assets/fields.json',
                  recipients: '/assets/recipients.json',
                  labels: '/assets/labels.json',
                  languages: '/assets/languages.json',
                  channels: '/assets/channels.json',
                  environment: '/assets/environment.json',
                  activity: '',
                  simulateStart: '',
                  simulateResume: ''
              }
          });
