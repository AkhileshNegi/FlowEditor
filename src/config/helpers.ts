import { Operator, Type } from '~/config';
import { FlowTypes, FlowTypeVisibility } from '~/config/interfaces';

export const filterOperators = (operators: Operator[], flowType: FlowTypes): Operator[] => {
    // return filterVisibility(operators, flowType);
    return operators;
};

export const filterTypeConfigs = (typeConfigs: Type[], flowType: FlowTypes): Type[] => {
    return filterVisibility(typeConfigs, flowType);
};

const filterVisibility = (items: FlowTypeVisibility[], flowType: FlowTypes): any[] => {
    return items.filter((item: FlowTypeVisibility) => {
        if (item.visibility === undefined) {
            return true;
        }

        return item.visibility.findIndex((ft: FlowTypes) => ft === flowType) > -1;
    });
};
