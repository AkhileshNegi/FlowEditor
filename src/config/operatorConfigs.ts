import { HIDDEN, Operator, OperatorMap, Operators, TEXT_TYPES } from '~/config/interfaces';

export const operatorConfigList: Operator[] = [
    {
        type: Operators.has_any_word,
        verboseName: 'has any of the words',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_all_words,
        verboseName: 'has all of the words',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_phrase,
        verboseName: 'has the phrase',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_only_phrase,
        verboseName: 'has only the phrase',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_beginning,
        verboseName: 'starts with',
        operands: 1
    },
    {
        type: Operators.has_text,
        verboseName: 'has some text',
        operands: 0,
        categoryName: 'Has Text',
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_number,
        verboseName: 'has a number',
        operands: 0,
        categoryName: 'Has Number'
    },
    {
        type: Operators.has_number_between,
        verboseName: 'has a number between',
        operands: 2
    },
    {
        type: Operators.has_number_lt,
        verboseName: 'has a number below',
        operands: 1
    },
    {
        type: Operators.has_number_lte,
        verboseName: 'has a number at or below',
        operands: 1
    },
    {
        type: Operators.has_number_eq,
        verboseName: 'has a number equal to',
        operands: 1
    },
    {
        type: Operators.has_number_gte,
        verboseName: 'has a number at or above',
        operands: 1
    },
    {
        type: Operators.has_number_gt,
        verboseName: 'has a number above',
        operands: 1
    },
    {
        type: Operators.has_date,
        verboseName: 'has a date',
        operands: 0,
        categoryName: 'Has Date',
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_date_lt,
        verboseName: 'has a date before',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_date_eq,
        verboseName: 'has a date equal to',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_date_gt,
        verboseName: 'has a date after',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_run_status,
        verboseName: 'has a run status of',
        operands: 1,
        visibility: HIDDEN
    },
    {
        type: Operators.has_group,
        verboseName: 'is in the group',
        operands: 1,
        visibility: HIDDEN
    },
    {
        type: Operators.has_phone,
        verboseName: 'has a phone number',
        operands: 0,
        categoryName: 'Has Phone'
    },
    {
        type: Operators.has_email,
        verboseName: 'has an email',
        operands: 0,
        categoryName: 'Has Email',
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_error,
        verboseName: 'has an error',
        operands: 0,
        categoryName: 'Has Error',
        visibility: HIDDEN
    },
    {
        type: Operators.has_value,
        verboseName: 'is not empty',
        operands: 0,
        categoryName: 'Not Empty'
    },
    {
        type: Operators.has_wait_timed_out,
        verboseName: null,
        operands: 0,
        visibility: HIDDEN
    },
    {
        type: Operators.has_pattern,
        verboseName: 'matches regex',
        operands: 1
    }
];

export const operatorConfigMap: OperatorMap = operatorConfigList.reduce(
    (map: OperatorMap, operatorConfig: Operator) => {
        map[operatorConfig.type] = operatorConfig;
        return map;
    },
    {}
);

/**
 * Shortcut for constant lookup of operator config in operator configs map
 * @param {string} type - The type of the operator config to return, e.g. 'send_msg'
 * @returns {Object} - The operator config found at operatorConfigs[type] or -1
 */
export const getOperatorConfig = (type: Operators): Operator => operatorConfigMap[type];
