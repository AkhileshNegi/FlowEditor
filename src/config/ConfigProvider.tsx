import * as React from 'react';
import { Endpoints, FlowEditorConfig } from '~/flowTypes';

export const fakePropType: any = (): any => null;
fakePropType.isRequired = (): any => null;

interface ConfigProviderProps {
    config: FlowEditorConfig;
    children: any;
}

export interface ConfigProviderContext {
    endpoints: Endpoints;
    flow: string;
    debug: boolean;
    showDownload: boolean;
}

// ----------------------------------------------------------------------------------------------

export const SINGLE_CHILD_ERROR = 'ConfigProvider expects only one child component.';
export const VALID_CHILD_ERROR =
    'ConfigProvider expects a valid React element: https://reactjs.org/docs/react-api.html#isvalidelement';

export default class ConfigProvider extends React.Component<ConfigProviderProps> {
    public static childContextTypes = {
        assetService: fakePropType,
        endpoints: fakePropType,
        flow: fakePropType,
        debug: fakePropType,
        showDownload: fakePropType
    };

    constructor(props: ConfigProviderProps) {
        super(props);

        if (React.Children.count(props.children) > 1) {
            throw new Error(SINGLE_CHILD_ERROR);
        } else if (!React.isValidElement(props.children)) {
            throw new Error(VALID_CHILD_ERROR);
        }
    }

    public getChildContext(): ConfigProviderContext {
        return {
            endpoints: this.props.config.endpoints,
            flow: this.props.config.flow,
            debug: this.props.config.debug,
            showDownload: this.props.config.showDownload
        };
    }

    public render(): JSX.Element {
        return React.Children.only(this.props.children);
    }
}
