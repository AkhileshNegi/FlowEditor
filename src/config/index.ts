import ConfigProvider, { ConfigProviderContext } from './ConfigProvider';
import { Type, Mode, typeConfigList, getTypeConfig } from './typeConfigs';
import { Operator, operatorConfigList, getOperatorConfig } from './operatorConfigs';

export {
    ConfigProviderContext,
    Type,
    Mode,
    Operator,
    typeConfigList,
    getTypeConfig,
    operatorConfigList,
    getOperatorConfig
};

export default ConfigProvider;
