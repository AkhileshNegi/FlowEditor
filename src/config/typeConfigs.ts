import AddLabelsComp from '../component/actions/AddLabels/AddLabels';
import AddLabelsForm from '../component/actions/AddLabels/AddLabelsForm';
import CallWebhookComp from '../component/actions/CallWebhook/CallWebhook';
import AddGroupsForm from '../component/actions/ChangeGroups/AddGroupsForm';
import ChangeGroupsComp from '../component/actions/ChangeGroups/ChangeGroups';
import RemoveGroupsForm from '../component/actions/ChangeGroups/RemoveGroupsForm';
import MissingComp from '../component/actions/Missing/Missing';
import SendBroadcastComp from '../component/actions/SendBroadcast/SendBroadcast';
import SendBroadcastForm from '../component/actions/SendBroadcast/SendBroadcastForm';
import { SendBroadcastFormHelper } from '../component/actions/SendBroadcast/SendBroadcastFormHelper';
import SendEmailComp from '../component/actions/SendEmail/SendEmail';
import SendEmailForm from '../component/actions/SendEmail/SendEmailForm';
import SendMsgComp from '../component/actions/SendMsg/SendMsg';
import SendMsgForm from '../component/actions/SendMsg/SendMsgForm';
import SetContactAttrib from '../component/actions/SetContactAttrib/SetContactAttrib';
import SetContactAttribForm from '../component/actions/SetContactAttrib/SetContactAttribForm';
import { SetContactAttribFormHelper } from '../component/actions/SetContactAttrib/SetContactAttribFormHelper';
import SetRunResultComp from '../component/actions/SetRunResult/SetRunResult';
import SetRunResultForm from '../component/actions/SetRunResult/SetRunResultForm';
import StartFlowComp from '../component/actions/StartFlow/StartFlow';
import StartSessionComp from '../component/actions/StartSession/StartSession';
import StartSessionForm from '../component/actions/StartSession/StartSessionForm';
import { StartSessionFormHelper } from '../component/actions/StartSession/StartSessionFormHelper';
import GroupsRouter from '../component/routers/GroupsRouter';
import SubflowRouter from '../component/routers/SubflowRouter';
import SwitchRouter from '../component/routers/SwitchRouter';
import WebhookRouter from '../component/routers/WebhookRouter';
import { AnyAction, RouterTypes, UINodeTypes } from '../flowTypes';
import { NodeEditorForm } from '../store/nodeEditor';

/*
Old name	                New name	                Event(s) generated

add_urn	                    add_contact_urn	            contact_urn_added
add_to_group	            add_contact_groups	        contact_groups_added
remove_from_group	        remove_contact_groups	    contact_groups_removed
set_preferred_channel	    set_contact_channel	        contact_channel_changed
update_contact	            set_contact_name	        contact_name_changed
update_contact	            set_contact_language	    contact_language_changed
update_contact	            set_contact_timezone	    contact_timezone_changed
save_contact_field      	set_contact_field	        contact_field_changed
save_flow_result	        set_run_result	            run_result_changed
call_webhook	            call_webhook	            webhook_called
add_label	                add_input_labels	        input_labels_added
reply	                    send_msg	                msg_created
send_email	                send_email	                email_created / email_sent
send_msg	                send_broadcast	            broadcast_created
start_flow	                start_flow	                flow_triggered
start_session	            start_session	            session_triggered
*/

export const enum Types {
    add_contact_urn = 'add_contact_urn',
    add_contact_groups = 'add_contact_groups',
    add_input_labels = 'add_input_labels',
    remove_contact_groups = 'remove_contact_groups',
    set_contact_channel = 'set_contact_channel',
    set_contact_name = 'set_contact_name',
    set_contact_field = 'set_contact_field',
    set_run_result = 'set_run_result',
    call_webhook = 'call_webhook',
    send_msg = 'send_msg',
    send_email = 'send_email',
    send_broadcast = 'send_broadcast',
    start_flow = 'start_flow',
    start_session = 'start_session',
    split_by_expression = 'split_by_expression',
    split_by_groups = 'split_by_groups',
    wait_for_response = 'wait_for_response',

    missing = 'missing'
}

export enum Mode {
    EDITING = 0x1,
    TRANSLATING = 0x2,
    ALL = EDITING | TRANSLATING
}

export interface FormHelper {
    actionToState: (action: AnyAction, actionType?: Types) => NodeEditorForm;
    stateToAction: (actionUUID: string, formState: NodeEditorForm, formType?: Types) => AnyAction;
}

export interface Type {
    type: Types;
    name: string;
    description: string;
    allows(mode: Mode): boolean;
    component?: React.SFC<AnyAction>;
    form?: React.ComponentClass<any>;
    formHelper?: FormHelper;
    advanced?: Mode;
    aliases?: string[];
}

export interface TypeMap {
    [propName: string]: Type;
}

export type GetTypeConfig = (type: string) => Type;

export function allows(mode: Mode): boolean {
    return (this.advanced & mode) === mode;
}

const ContactAttribHelper = new SetContactAttribFormHelper();

export const typeConfigList: Type[] = [
    /** Actions */
    {
        type: Types.missing,
        name: 'Missing',
        description: ' ** Unsupported ** ',
        component: MissingComp,
        allows
    },
    {
        type: Types.send_msg,
        name: 'Send Message',
        description: 'Send the contact a message',
        form: SendMsgForm,
        component: SendMsgComp,
        advanced: Mode.EDITING,
        allows
    },
    {
        type: Types.send_broadcast,
        name: 'Send Broadcast',
        description: 'Send somebody else a message',
        form: SendBroadcastForm,
        formHelper: new SendBroadcastFormHelper(),
        component: SendBroadcastComp,
        allows
    },
    {
        type: Types.add_input_labels,
        name: 'Add Labels',
        description: 'Label the incoming message',
        form: AddLabelsForm,
        component: AddLabelsComp,
        allows
    },
    {
        type: Types.add_contact_groups,
        name: 'Add to Group',
        description: 'Add the contact to a group',
        form: AddGroupsForm,
        component: ChangeGroupsComp,
        allows
    },
    {
        type: Types.remove_contact_groups,
        name: 'Remove from Group',
        description: 'Remove the contact from a group',
        form: RemoveGroupsForm,
        component: ChangeGroupsComp,
        allows
    },
    {
        type: Types.set_contact_field,
        name: 'Update Contact',
        description: 'Update the contact',
        form: SetContactAttribForm,
        formHelper: ContactAttribHelper,
        component: SetContactAttrib,
        allows
    },
    {
        type: Types.set_contact_name,
        name: 'Update Contact',
        description: 'Update the contact',
        form: SetContactAttribForm,
        formHelper: ContactAttribHelper,
        component: SetContactAttrib,
        allows
    },
    {
        type: Types.send_email,
        name: 'Send Email',
        description: 'Send an email',
        form: SendEmailForm,
        component: SendEmailComp,
        allows
    },
    {
        type: Types.set_run_result,
        name: 'Save Flow Result',
        description: 'Save a result for this flow',
        form: SetRunResultForm,
        component: SetRunResultComp,
        allows
    },
    // {type: 'set_preferred_channel', name: 'Set Preferred Channel', description: 'Set their preferred channel', component: Missing},
    /** Hybrids */
    {
        type: Types.call_webhook,
        name: 'Call Webhook',
        description: 'Call a webook',
        form: WebhookRouter,
        component: CallWebhookComp,
        advanced: Mode.EDITING,
        aliases: [UINodeTypes.webhook],
        allows
    },
    {
        type: Types.start_flow,
        name: 'Start a Flow',
        description: 'Enter another flow',
        form: SubflowRouter,
        component: StartFlowComp,
        aliases: [UINodeTypes.subflow],
        allows
    },
    {
        type: Types.start_session,
        name: 'Start Somebody Else',
        description: 'Start somebody else in a flow',
        form: StartSessionForm,
        formHelper: new StartSessionFormHelper(),
        component: StartSessionComp,
        allows
    },

    /** Routers */
    {
        type: Types.split_by_expression,
        name: 'Split by Expression',
        description: 'Split by a custom expression',
        form: SwitchRouter,
        advanced: Mode.TRANSLATING,
        allows
    },
    {
        type: Types.split_by_groups,
        name: 'Split by Group Membership',
        description: 'Split by group membership',
        form: GroupsRouter,
        allows
    },
    {
        type: Types.wait_for_response,
        name: 'Wait for Response',
        description: 'Wait for the contact to respond',
        form: SwitchRouter,
        advanced: Mode.TRANSLATING,
        aliases: [RouterTypes.switch],
        allows
    }
    // {type: 'random', name: 'Random Split', description: 'Split them up randomly', form: RandomRouterForm}
];

export const typeConfigMap: TypeMap = typeConfigList.reduce((map: TypeMap, typeConfig: Type) => {
    map[typeConfig.type] = typeConfig;
    if (typeConfig.aliases) {
        typeConfig.aliases.forEach((alias: string) => (map[alias] = typeConfig));
    }
    return map;
}, {});

/**
 * Shortcut for constant lookup of type config in type configs map
 * @param {string} type - The type of the type config to return, e.g. 'send_msg'
 * @returns {Object} - The type config found at typeConfigs[type] or -1
 */
export const getTypeConfig = (type: Types | RouterTypes | UINodeTypes): Type => {
    let actionConfig = typeConfigMap[type];

    if (!actionConfig) {
        actionConfig = typeConfigMap.missing;
    }
    return actionConfig;
};
