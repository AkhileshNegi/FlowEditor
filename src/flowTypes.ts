import {
    IsOptionUniqueHandler,
    IsValidNewOptionHandler,
    NewOptionCreatorHandler,
    PromptTextCreatorHandler
} from 'react-select';
import { Operators } from './config/operatorConfigs';
import AssetService from './services/AssetService';
import { Types } from './config/typeConfigs';

export interface Languages {
    [iso: string]: string;
}

export interface Endpoints {
    fields: string;
    groups: string;
    contacts: string;
    flows: string;
    activity: string;
    labels: string;
    simulateStart: string;
    simulateResume: string;
}

export interface FlowEditorConfig {
    assetService?: AssetService;
    localStorage: boolean;
    languages: { [iso: string]: string };
    endpoints: Endpoints;
    flow: string;
    path?: string;
}

export interface LocalizationMap {
    [lang: string]: {
        [uuid: string]: any;
    };
}

export interface FlowDefinition {
    localization: LocalizationMap;
    language: string;
    name: string;
    nodes: FlowNode[];
    uuid: string;
    _ui: UIMetaData;
}

export interface FlowNode {
    uuid: string;
    actions: Action[];
    exits: Exit[];
    router?: Router;
    wait?: Wait;
}

export interface Exit {
    uuid: string;
    name?: string;
    destination_node_uuid?: string;
}

export enum RouterTypes {
    switch = 'switch'
}

export interface Router {
    type: RouterTypes;
    result_name?: string;
}

export interface Case {
    uuid: string;
    type: Operators;
    exit_uuid: string;
    arguments?: string[];
}

export interface SwitchRouter extends Router {
    cases: Case[];
    operand: string;
    default_exit_uuid: string;
}

export enum WaitTypes {
    exp = 'exp',
    group = 'group',
    msg = 'msg',
    flow = 'flow'
}

export interface Wait {
    type: WaitTypes;
    timeout?: number;
}

export interface Group {
    uuid: string;
    name: string;
}

export enum Methods {
    GET = 'GET',
    POST = 'POST',
    PUT = 'PUT'
}

export interface ChangeGroups extends Action {
    groups: Group[];
}

export interface Field {
    key: string;
    name: string;
}

export interface Label {
    uuid: string;
    name: string;
}

export interface Action {
    type: Types;
    uuid: string;
}

export interface SetContactField extends Action {
    field: Field;
    value: string;
}

export interface SetContactName extends Action {
    type: Types.set_contact_name;
    name: string;
}

export type SetContactProperty = SetContactName;

export type SetContactAttribute = SetContactField | SetContactProperty;

// tslint:disable-next-line:no-empty-interface
export interface Missing extends Action {}

export interface SendMsg extends Action {
    text: string;
    all_urns?: boolean;
}

export interface AddLabels extends Action {
    labels: Label[];
}

export interface SetPreferredChannel extends Action {
    language: string;
}

export interface SendEmail extends Action {
    subject: string;
    body: string;
    addresses: string[];
}

export interface SetRunResult extends Action {
    name: string;
    value: string;
    category?: string;
}

export interface Headers {
    [name: string]: string;
}

export interface CallWebhook extends Action {
    url: string;
    method: Methods;
    body?: string;
    headers?: Headers;
}

export interface StartFlow extends Action {
    flow: {
        name: string;
        uuid: string;
    };
}

export interface UIMetaData {
    nodes: { [key: string]: UINode };
    languages: Array<{ [iso: string]: string }>;
}

export interface FlowPosition {
    left: number;
    top: number;
    right?: number;
    bottom?: number;
}

export interface Dimensions {
    width: number;
    height: number;
}

export enum UINodeTypes {
    split = 'split',
    subflow = 'subflow',
    webhook = 'webhook'
}

export interface UINode {
    position: FlowPosition;
    // ui type, used for split by expression, contact field, etc
    type?: UINodeTypes | Types;
}

export interface StickyNote {
    position: FlowPosition;
    title: string;
    body: string;
    color?: string;
}

export interface UIMetaData {
    nodes: { [key: string]: UINode };
    stickies: { [key: string]: StickyNote };
}

export type AnyAction =
    | Action
    | ChangeGroups
    | SetContactField
    | SetContactProperty
    | SetRunResult
    | SendMsg
    | SetPreferredChannel
    | SendEmail
    | CallWebhook
    | StartFlow;

export enum ContactProperties {
    UUID = 'UUID',
    'Created By' = 'Created By',
    'Modified By' = 'Modified By',
    Org = 'Org',
    Name = 'Name',
    Language = 'Language',
    Timezone = 'Timezone',
    Email = 'Email',
    Mailto = 'Mailto',
    Phone = 'Phone',
    Groups = 'Groups',
    Facebook = 'Facebook',
    Telegram = 'Telegram'
}

export enum ResultType {
    flow = 'flow',
    field = 'field',
    group = 'group',
    label = 'label'
}

export enum ValueType {
    text = 'text',
    numeric = 'numeric',
    datetime = 'datetime',
    state = 'state',
    district = 'district',
    ward = 'ward'
}

export interface CreateOptions {
    promptTextCreator?: PromptTextCreatorHandler;
    newOptionCreator?: NewOptionCreatorHandler;
    isValidNewOption?: IsValidNewOptionHandler;
    isOptionUnique?: IsOptionUniqueHandler;
    createNewOption?: NewOptionCreatorHandler;
    createPrompt?: string;
}

export enum StartFlowArgs {
    Complete = 'C',
    Expired = 'E'
}

export enum StartFlowExitNames {
    Complete = 'Complete',
    Expired = 'Expired'
}

export enum WebhookExitNames {
    Success = 'Success',
    Failure = 'Failed',
    Unreachable = 'Unreachable'
}
