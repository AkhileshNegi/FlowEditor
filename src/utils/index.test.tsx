import * as React from 'react';
import {
    addCommas,
    emphasize,
    getBaseLanguage,
    getLanguage,
    getLocalization,
    getSelectClass,
    hasErrorType,
    jsonEqual,
    reorderList,
    snakify,
    titleCase,
    toBoolMap,
    validUUID,
    propertyExists,
    renderIf
} from '.';
import { operatorConfigList } from '../config';
import { ContactProperties } from '../flowTypes';

const config = require('../../assets/config');
const {
    results: [{ definition: colorsFlow }]
} = require('../../assets/flows/a4f64f1b-85bc-477e-b706-de313a022979.json');
const { nodes: [{ actions: [sendMsgAction] }] } = colorsFlow;

describe('utils', () => {
    describe('toBoolMap', () => {
        it('should turn a string array into a bool map', () => {
            const strArr: string[] = [
                'language',
                'facebook',
                'telegram',
                'email',
                'mailto',
                'name',
                'first name',
                'phone',
                'groups',
                'uuid',
                'created by',
                'modified by',
                'org',
                'is',
                'has',
                'tel'
            ];
            const expected: { [key: string]: boolean } = {
                'created by': true,
                email: true,
                facebook: true,
                'first name': true,
                groups: true,
                has: true,
                is: true,
                language: true,
                mailto: true,
                'modified by': true,
                name: true,
                org: true,
                phone: true,
                tel: true,
                telegram: true,
                uuid: true
            };

            expect(toBoolMap(strArr)).toEqual(expected);
        });
    });

    describe('addCommas', () => {
        it('should insert commas into numbers where appropriate', () => {
            expect(addCommas(999)).toBe('999');
            expect(addCommas(10000)).toBe('10,000');
            expect(addCommas(10000000)).toBe('10,000,000');
        });
    });

    describe('snakify', () => {
        it('should replace spaces with underscores', () => {
            expect(snakify('my result name')).toBe('my_result_name');
        });
    });

    describe('validUUID', () => {
        it('matches a valid v4 UUID', () => {
            const v4 = '9ddd2483-3071-498d-bd4e-0fe9c0b2fa94';
            const v4Upper = '06FDAF8D-9905-4DA7-B94D-7BA3532D3953';

            expect(validUUID(v4)).toBeTruthy();
            expect(validUUID(v4Upper)).toBeTruthy();
        });

        it("doesn't match an invalid v4 UUID", () => {
            const v1 = '63955E30-B5DE-11E7-8F1A-0800200C9A66';
            const randStr = 'ZUBikEKpzoD7XOdo74Ux';

            expect(validUUID(v1)).toBeFalsy();
            expect(validUUID(randStr)).toBeFalsy();
        });
    });

    describe('getSelectClass', () => {
        it('should return an empty array if passed "errors" arg less than 1', () =>
            expect(getSelectClass(0)).toEqual(''));

        it("should return an array containing react-select's invalid class", () =>
            expect(getSelectClass(1)).toEqual('select-invalid'));
    });

    describe('titleCase', () =>
        it('should apply title case to each word in a given string', () =>
            ['one', 'one two', 'one two three', 'one,two', 'one, two'].forEach(str =>
                expect(titleCase(str)).toMatchSnapshot()
            )));

    describe('reorderList', () =>
        it('should reorder a given list according to index params', () =>
            expect(reorderList([1, 2, 3], 2, 0)).toEqual([3, 1, 2])));

    describe('jsonEqual', () => {
        const [anyWordOperator, allWordsOperator] = operatorConfigList;

        it('should return true if basic objects are equal in contents and order', () =>
            expect(jsonEqual(anyWordOperator, anyWordOperator)).toBeTruthy());

        it('should return false if basic objects are not equal in contents and order', () =>
            expect(jsonEqual(anyWordOperator, allWordsOperator)).toBeFalsy());
    });

    describe('hasErrorType', () => {
        it('should return false if passed an empty error list', () =>
            expect(hasErrorType([], [/argument/])).toBeFalsy());

        it('should return true if query exits in a string in the error list', () => {
            const errors = ['A category name is required.'];

            expect(hasErrorType(errors, [/category/])).toBeTruthy();
        });
    });

    describe('getBaseLanguage', () => {
        it('should return the language stored in the first key of a Languages map', () => {
            const baseLanguage = getBaseLanguage(config.languages);
            const firstKey = Object.keys(config.languages)[0];

            expect(baseLanguage.iso).toBe(firstKey);
            expect(baseLanguage.name).toBe(config.languages[firstKey]);
        });
    });

    describe('getLanguage', () => {
        it('should return specified language from Languages map if it exists in map', () => {
            const iso = Object.keys(config.languages)[0];

            expect(getLanguage(config.languages, iso)).toEqual({
                name: config.languages[iso],
                iso
            });
        });
    });

    describe('getLocalizations', () => {
        it('should return a localized object', () => {
            ['eng', 'spa', 'fre'].forEach(iso => {
                expect(
                    getLocalization(sendMsgAction, colorsFlow.localization, iso, config.languages)
                ).toMatchSnapshot();
            });
        });
    });

    describe('getLanguage', () => {
        it('should return language as Language', () => {
            Object.keys(config.languages).forEach(iso => {
                const language = getLanguage(config.languages, iso);

                expect(language).toEqual({
                    name: config.languages[iso],
                    iso
                });
                expect(language).toMatchSnapshot();
            });
        });
    });

    describe('getBaseLanguage', () => {
        it("should return Language corresponding to the first property in the Languages object it's passed", () => {
            const baseLanguage = getBaseLanguage(config.languages);

            expect(baseLanguage).toMatchSnapshot();
        });
    });

    describe('emphasize', () => {
        it('should apply emphasis style', () => {
            expect(emphasize('emboldened')).toMatchSnapshot();
        });
    });

    describe('propertyExists', () => {
        it('should return true if property exists on ContactProperties enum', () => {
            Object.keys(ContactProperties).forEach(property =>
                expect(propertyExists(property)).toBeTruthy()
            );
        });

        it('should return false if property does not exist on ContactProperties enum', () => {
            expect(propertyExists('Favorite Song')).toBeFalsy();
        });
    });

    describe('renderIf', () => {
        it('should return element if condition is truthy', () => {
            expect(renderIf(true)(<div />)).toEqual(<div />);
        });

        it('should return null if condition is falsy', () => {
            expect(renderIf(false)(<div />)).toBeNull();
        });
    });
});
