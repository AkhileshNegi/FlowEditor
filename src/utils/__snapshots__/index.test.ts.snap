// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils emphasize should apply emphasis style 1`] = `
<span
  className="emph"
>
  emboldened
</span>
`;

exports[`utils titleCase should apply title case to each word in a given string 1`] = `"One"`;

exports[`utils titleCase should apply title case to each word in a given string 2`] = `"One Two"`;

exports[`utils titleCase should apply title case to each word in a given string 3`] = `"One Two Three"`;

exports[`utils titleCase should apply title case to each word in a given string 4`] = `"One,Two"`;

exports[`utils titleCase should apply title case to each word in a given string 5`] = `"One, Two"`;
